src/pebble.py

```diff
@@ -121,36 +224,36 @@ def change_config(charm_state: CharmState, container: ops.model.Container) -> No
     """
     try:
         synapse.execute_migrate_config(container=container, charm_state=charm_state)
-        synapse.enable_metrics(container=container)
-        synapse.enable_forgotten_room_retention(container=container)
-        synapse.enable_serve_server_wellknown(container=container)
+        current_yaml = _get_synapse_config(container)
+        synapse.enable_metrics(current_yaml)
+        synapse.enable_forgotten_room_retention(current_yaml)
+        synapse.enable_serve_server_wellknown(current_yaml)
         if charm_state.saml_config is not None:
             logger.debug("pebble.change_config: Enabling SAML")
-            synapse.enable_saml(container=container, charm_state=charm_state)
+            synapse.enable_saml(current_yaml, charm_state=charm_state)
         if charm_state.smtp_config is not None:
             logger.debug("pebble.change_config: Enabling SMTP")
-            synapse.enable_smtp(container=container, charm_state=charm_state)
+            synapse.enable_smtp(current_yaml, charm_state=charm_state)
         if charm_state.redis_config is not None:
             logger.debug("pebble.change_config: Enabling Redis")
-            synapse.enable_redis(container=container, charm_state=charm_state)
+            synapse.enable_redis(current_yaml, charm_state=charm_state)
         if not charm_state.synapse_config.enable_password_config:
-            synapse.disable_password_config(container=container)
+            synapse.disable_password_config(current_yaml)
         if charm_state.synapse_config.federation_domain_whitelist:
-            synapse.enable_federation_domain_whitelist(
-                container=container, charm_state=charm_state
-            )
+            synapse.enable_federation_domain_whitelist(current_yaml, charm_state=charm_state)
         if charm_state.synapse_config.allow_public_rooms_over_federation:
-            synapse.enable_allow_public_rooms_over_federation(container=container)
+            synapse.enable_allow_public_rooms_over_federation(current_yaml)
         if not charm_state.synapse_config.enable_room_list_search:
-            synapse.disable_room_list_search(container=container)
+            synapse.disable_room_list_search(current_yaml)
         if charm_state.synapse_config.trusted_key_servers:
-            synapse.enable_trusted_key_servers(container=container, charm_state=charm_state)
+            synapse.enable_trusted_key_servers(current_yaml, charm_state=charm_state)
         if charm_state.synapse_config.ip_range_whitelist:
-            synapse.enable_ip_range_whitelist(container=container, charm_state=charm_state)
+            synapse.enable_ip_range_whitelist(current_yaml, charm_state=charm_state)
         if charm_state.datasource:
             logger.info("Synapse Stats Exporter enabled.")
             replan_stats_exporter(container=container, charm_state=charm_state)
-        synapse.validate_config(container=container)
+        synapse.validate_config(container)
```

Comment:
The yaml should be updated in the container before running validate_config

Reply:
Or could be refactored to run in a temporary file, it would work too.

Reply:
Good catch! I moved it after pushing (for now).